test/test_sparse.py

```diff
@@ -5,227 +5,249 @@
 import random
 import unittest
 from common import TestCase, run_tests
+from common_nn import TEST_CUDA
 from numbers import Number
 
-SparseTensor = sparse.DoubleTensor
+# triplet := (index type, value type, sparse type)
+cpu_triplet = (
+    torch.LongTensor,
+    torch.DoubleTensor,
+    torch.sparse.DoubleTensor)
+type_triplets = [cpu_triplet]
+if TEST_CUDA:
+    cuda_triplet = (
+        torch.cuda.LongTensor,
+        torch.cuda.DoubleTensor,
+        torch.cuda.sparse.DoubleTensor)
+    type_triplets.append(cuda_triplet)
 
 
 class TestSparse(TestCase):
 
     @staticmethod
-    def _gen_sparse(d, nnz, with_size):
+    def _gen_sparse(d, nnz, with_size, is_cuda=False):  # FIXME remove default is_cuda value to ensure coverage
         if isinstance(with_size, Number):
             v = torch.randn(nnz)
             i = (torch.rand(d, nnz) * with_size).type(torch.LongTensor)
-            x = SparseTensor(i, v)
+            x = torch.sparse.DoubleTensor(i, v)
         else:
             v_size = [nnz] + list(with_size[d:])
             v = torch.randn(*v_size)
             i = torch.rand(d, nnz) * \
                 torch.Tensor(with_size[:d]).repeat(nnz, 1).transpose(0, 1)
             i = i.type(torch.LongTensor)
-            x = SparseTensor(i, v, torch.Size(with_size))
+            x = torch.sparse.DoubleTensor(i, v, torch.Size(with_size))
 
-        return x, i, v
+        if is_cuda:
+            return x.cuda(), i.cuda(), v.cuda()
+        else:
+            return x, i, v
 
     def test_basic(self):
-        x, i, v = self._gen_sparse(3, 10, 100)
-
-        self.assertEqual(i, x.indices())
-        self.assertEqual(v, x.values())
-
-        x, i, v = self._gen_sparse(3, 10, [100, 100, 100])
-        self.assertEqual(i, x.indices())
-        self.assertEqual(v, x.values())
-        self.assertEqual(x.ndimension(), 3)
-        self.assertEqual(x.nnz(), 10)
-        for i in range(3):
-            self.assertEqual(x.size(i), 100)
-
-        # Make sure we can access empty indices / values
-        x = SparseTensor()
-        self.assertEqual(x.indices().numel(), 0)
-        self.assertEqual(x.values().numel(), 0)
+        for is_cuda in [False, True] if TEST_CUDA else [False]:
```

Comment:
can you separate CUDA and non-CUDA tests like we do in all other tests? Put code for this method in `_test_basic` and make it accept `is_cuda` argument. Then add `test_basic` and `test_basic_cuda`, and decorate the latter one with `@unittest.skipIf`. It's better because you can actually see that some tests were skipped, so they might be incomplete.

Same applies to all `type_triplets` tests. Make the helper accept `type_triplet` arg and give it `cpu_type_triplet` and `gpu_type_triplet` in the corresponding test methods.